pkg/policy/selectorcache.go

```diff
@@ -238,6 +244,9 @@ func (s *selectorManager) Equal(b *selectorManager) bool {
 // of the selections. If the old version is returned, the user is
 // guaranteed to receive a notification including the update.
 func (s *selectorManager) GetSelections() []identity.NumericIdentity {
+	if s.selections == nil {
+		return []identity.NumericIdentity{}
+	}
```

Comment:
The fact that this is needed here is an indication that we somewhere add cached selections to the cache that are not properly initialized. After checking all the call sites that add cached selections to the cache, I think it happens in  `updateFQDNSelector()`. It can add a selector to the cache that has no user (which is problematic in itself), but it also does not call the `updateSelections()`, which would properly initialize `selections` so that the `nil` check here would not be needed.

So, instead of adding a `nil` check here, IMO you should figure out how to fix `updateFQDNSelector()`. Adding the `updateSelections()` call is a trivial fix. But if `updateFQDNSelector()` can insert a new cached selection, it would also take a `user` parameter and initialize the first user just like `AddFQDNSelector()` does. Furthermore, it would be good to understand why `updateFQDNSelector()` needs to insert a new cached selector to begin with? IMO we should be able to rely on the `AddFQDNSelector()` calls and simply ignore non-found selectors in `updateFQDNSelector()` as not finding the selector from the cache should be an indication that the current policy is not interested in that given selector?

Reply:
Going back with a fresh set of eyes, I think I was a bit wrong above:

`updateFQDNSelector()` *does* call `updateSelections()`, but after the new cached selector has been placed in the map. While this is different compared to the other call sites, this is not wrong, as the whole operation is protected by the lock, and the map lookups take the same lock.

A `GetSelections()` call can only be made after the cached (FQDN) selection is returned from the `AddFQDNSelector()` call, so I don't find any way a `GetSelections()` call could be made on a cached selection where `s.selections == nil`??

Reply:
This may be an artifact left over from my old way of adding stuff into the cache. I'll remove it and trigger another CI run.